### DISCUSSION TITLE
"App not installed as package apprears to be invalid " on ionic cordova Android .apk

### DISCUSSION THREAD

**Discussion by jis248** [codemagic-ci-cd/codemagic-docs#2272] [Q&A]
### If you are looking for help, please confirm the following...

- [X] I confirm I have searched the [Docs](https://docs.codemagic.io), [Codemagic Sample Projects](https://github.com/codemagic-ci-cd/codemagic-sample-projects), and GitHub [Discussions](https://github.com/orgs/codemagic-ci-cd/discussions).
- [X] I confirm the software versions on my local computer match the ones in CI environment (e.g. Xcode version, Gradle version). See [here](https://docs.codemagic.io/troubleshooting/common-android-issues/#builds-work-locally-but-fail-on-codemagic)
- [X] I confirm I am able to do a fresh clone of the repository and build the project on my local computer.

### Which mobile framework are you using?

Ionic/Cordova

### Steps to reproduce

Build is a Success and after install on android it gives error
"App not installed as package apprears to be invalid"

// codemagic.yaml

 android-workflow:
    name: Android workflow
    environment:
      android_signing:
        - keystore_android
      # groups:
      #   - google_play # <-- (Includes GCLOUD_SERVICE_ACCOUNT_CREDENTIALS - Put your google-services.json here)
      java: 1.8
      node: v16.14.0
      npm: 9.1.2
    cache:
      cache_paths:
        - $CM_BUILD_DIR/node_modules
    scripts:
      - name: Install dependencies
        script: |
          wget https://services.gradle.org/distributions/gradle-5.6.2-bin.zip
          unzip -q gradle-5.6.2-bin.zip
          export GRADLE_HOME=$PWD/gradle-5.6.2
          export PATH=$GRADLE_HOME/bin:$PATH
          gradle -v
          npm install
          npm ci # equivalent of npm install for CI systems.
            # Requires package-lock.json or npm-shrinkwrap.json to be present
          cvm install 9.0.0
          cvm use 9.0.0
          npm install -g ionic
          ionic info
      - name: Add Android platform
        script: |
          ionic cordova platform remove android --nosave
          ionic cordova platform add android \
            --confirm \
            --no-interactive \
            --noresources
      - name: Build Android
        script: |
          ionic cordova build android \
           --release \
           --no-interactive \
           --prod \
           --device
      - name: Sign APK
        script: |
          APK_PATH=$(find platforms/android/app/build/outputs/apk/release -name "*.apk" | head -1)
          jarsigner \
            -sigalg SHA256withRSA \
            -digestalg SHA-256 \
            -keystore $CM_KEYSTORE_PATH \
            -storepass $CM_KEYSTORE_PASSWORD \
            -keypass $CM_KEY_PASSWORD \
            $APK_PATH $CM_KEY_ALIAS
      - name: Verify Sign APK
        script: |
          APK_PATH=$(find platforms/android/app/build/outputs/apk/release -name "*.apk" | head -1)
          jarsigner \
            -verify \
            -verbose \
            $APK_PATH $CM_KEY_ALIAS
    artifacts:
      - platforms/android/app/build/outputs/**/*.apk
      - platforms/android/app/build/outputs/**/mapping.txt

### Expected results

apk should work in all android devices

### Actual results

This happens for new versions of android but works on some older versions.
note: There is no sdk version issue.

i have created a release apk on my local and it works on all devices. I have checked the verify apk commands and  signed apk.

My system 
![image](https://github.com/codemagic-ci-cd/codemagic-docs/assets/84921389/43b1c949-10b0-42ea-8b96-86f1a6085fa1)

**Comment by icarusdust** [comment]
Hi @jis248, referring to your sentence, _**This happens for new versions of android but works on some older versions.**_, I am assuming that when building with Android 11, the package installs ok, but when installing a higher version, you are getting the aforementioned error. Though, I do not see any script that installs a higher Android SDK in your YAML. Anyways, I will need you to elaborate on it and please provide build IDs where it works and fails.

**Reply by jis248** [reply]
Build ID:- 657c5c0aada0c1a1e076af4c

Same Build works on android 11 but does not work on android 12 and above. No there is no different scripts used.

**Reply by icarusdust** [reply]
It sounds like a possible system settings or permissions issue that prevents the app from being installed correctly. Try to reset the app settings from your device settings and check if any permissions from the device need to be changed.

Other than that, device logs should print out the error message or some exceptions that could help you understand what the issue is about exactly.

**Comment by icarusdust** [comment]
I am closing the thread due to inactivity. Let me know if you have any more questions by replying to this thread or opening a new one.
